expression/builtin_regexp.go

```diff
@@ -0,0 +1,1322 @@
+		regexpBaseFuncSig: regexpBaseFuncSig{baseBuiltinFunc: bf},
+	}
+
+	if c.funcName == ast.Regexp {
+		sig.setPbCode(tipb.ScalarFuncSig_RegexpSig)
+	} else {
+		sig.setPbCode(tipb.ScalarFuncSig_RegexpLikeSig)
+	}
+
+	return &sig, nil
+}
+
+type builtinRegexpLikeFuncSig struct {
+	regexpBaseFuncSig
+}
+
+func (re *builtinRegexpLikeFuncSig) Clone() builtinFunc {
+	newSig := &builtinRegexpLikeFuncSig{}
+	newSig.cloneFrom(&re.baseBuiltinFunc)
+	newSig.clone(&re.regexpBaseFuncSig)
+	return newSig
+}
+
+func (re *builtinRegexpLikeFuncSig) vectorized() bool {
+	return true
+}
+
+func (re *builtinRegexpLikeFuncSig) evalInt(row chunk.Row) (int64, bool, error) {
+	expr, isNull, err := re.args[0].EvalString(re.ctx, row)
+	if isNull || err != nil {
+		return 0, true, err
+	}
+
+	pat, isNull, err := re.args[1].EvalString(re.ctx, row)
+	if isNull || err != nil {
+		return 0, true, err
+	}
+
+	matchType := ""
+	if len(re.args) == 3 {
+		matchType, isNull, err = re.args[2].EvalString(re.ctx, row)
+		if isNull || err != nil {
+			return 0, true, err
+		}
+	}
+
+	compile, err := re.genCompile(matchType)
+	if err != nil {
+		return 0, true, ErrRegexp.GenWithStackByArgs(err.Error())
+	}
+
+	reg, err := compile(pat)
+	if err != nil {
+		return 0, true, ErrRegexp.GenWithStackByArgs(err.Error())
+	}
+
+	return boolToInt64(reg.MatchString(expr)), false, nil
+}
+
+// we need to memorize the regexp when:
+//  1. pattern and match type are constant
+//  2. pattern is const and there is no match type argument
+//
+// return true: need, false: needless
+func (re *builtinRegexpLikeFuncSig) needMemorization() bool {
+	return (re.args[1].ConstItem(re.ctx.GetSessionVars().StmtCtx) && (len(re.args) <= 2 || re.args[2].ConstItem(re.ctx.GetSessionVars().StmtCtx))) && !re.isMemorizedRegexpInitialized()
+}
+
+// Call this function when at least one of the args is vector
+// REGEXP_LIKE(expr, pat[, match_type])
+func (re *builtinRegexpLikeFuncSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	params := make([]*regexpParam, 0, 3)
+	defer releaseBuffers(&re.baseBuiltinFunc, params)
+
+	for i := 0; i < 2; i++ {
+		param, isConstNull, err := buildStringParam(&re.baseBuiltinFunc, i, input, false)
+		if err != nil {
+			return err
+		}
+		if isConstNull {
+			result.ResizeInt64(n, true)
+			return nil
+		}
+		params = append(params, param)
+	}
+
+	// user may ignore match type parameter
+	hasMatchType := (len(re.args) == 3)
+	param, isConstNull, err := buildStringParam(&re.baseBuiltinFunc, 2, input, !hasMatchType)
+	params = append(params, param)
+
+	if err != nil {
+		return err
+	}
+
+	if isConstNull {
+		result.ResizeInt64(n, true)
+		return nil
+	}
+
+	// Check memorization
+	if re.needMemorization() {
+		// matchType must be const or null
+		matchType := params[2].getStringVal(0)
+
+		compile, err := re.genCompile(matchType)
+		if err != nil {
+			return err
+		}
+
+		re.initMemoizedRegexp(compile, params[1].getCol(), n)
```

Comment:
according to #16600, there might be some multi-thread issues, please double comfirm that `initMemoizedRegexp` need to be protected from multi-thread issues.